src/Chapter3.hs

```diff
@@ -1006,6 +1102,95 @@ Implement data types and typeclasses, describing such a battle between two
 contestants, and write a function that decides the outcome of a fight!
 -}
 
+data Monster = Monster {mHP :: Int, mDmg :: Int} deriving Show
+data Knight = Knight { kHP  :: Int, kDef :: Int, kDmg :: Int} deriving Show
+
+type Effect = forall c. Combatant c => c -> IO c
```

Comment:
Cool 😎 
I don't think that you need my advice as you managed such complicated topics already 👍🏼

Reply:
I'm not really confident with language extensions, just wanted to make single list of actions, dont know if this is acceptable way